src/Draco.Lsp/Server/LanguageServerConnection.cs

```diff
@@ -0,0 +1,604 @@
+                {
+                    buffer = buffer.Slice(reader.Position, reader.Position);
+
+                    var jsonReader = new Utf8JsonReader(utf8Json);
+                    message = JsonSerializer.Deserialize<LspMessage>(ref jsonReader, jsonOptions);
+                    return true;
+                }
+                else
+                {
+                    // We couldn't read the full JSON payload, so we'll come back here when more data is available.
+                    // Keep the \r\n right before the JSON payload. Otherwise, we'll never enter this loop again.
+                    reader.Rewind(2);
+                    buffer = buffer.Slice(reader.Position);
+                    break;
+                }
+            }
+
+            buffer = buffer.Slice(reader.Position);
+
+            var headerText = header.IsSingleSegment ? header.FirstSpan : header.ToArray();
+
+            var ContentLengthHeader = "Content-Length: "u8;
+            var ContentTypeHeader = "Content-Type: "u8;
+
+            if (headerText.StartsWith(ContentLengthHeader))
+            {
+                if (!Utf8Parser.TryParse(headerText[ContentLengthHeader.Length..], out contentLength, out _))
+                {
+                    throw new InvalidDataException();
+                }
+            }
+            else if (headerText.StartsWith(ContentTypeHeader))
+            {
+                // In theory, we are supposed to ensure that the requested encoding is UTF-8.
+            }
+        }
+
+        message = default;
+        return false;
+    }
+
+    private async Task SerializeToTransport(LspMessage message)
+    {
+        var writer = this.transport.Output;
+
+        void WriteData()
+        {
+            var response = JsonSerializer.SerializeToUtf8Bytes(message, jsonOptions);
+
+            var ContentLengthHeader = "Content-Length: "u8;
+            var TwoNewLines = "\r\n\r\n"u8;
+
+            var written = 0;
+            var responseBuffer = writer.GetSpan(response.Length + ContentLengthHeader.Length + 14); // 14 UTF-8 bytes can encode any integer and 4 newlines
+
+            ContentLengthHeader.CopyTo(responseBuffer);
+            written += ContentLengthHeader.Length;
+
+            Utf8Formatter.TryFormat(response.Length, responseBuffer[written..], out var contentLengthLength);
+            written += contentLengthLength;
+
+            TwoNewLines.CopyTo(responseBuffer[written..]);
+            written += TwoNewLines.Length;
+
+            response.CopyTo(responseBuffer[written..]);
+            written += response.Length;
+
+            writer.Advance(written);
+        }
+
+        WriteData();
+        var result = await writer.FlushAsync();
+        if (result.IsCompleted)
+        {
+            this.outgoingMessages.Complete();
+        }
+    }
+
+    private static readonly MethodInfo taskGetResult = typeof(Task<>).GetMethod("get_Result")!;
```

Comment:
Do we want to move this to the top?

Reply:
It's only used by the method below it. If C# had static local variables it would be one of those. I'm not sure if it's worth elevating it to the top where all of the real state is